### ISSUE TITLE
Code folding and command line execution

### ISSUE THREAD

**Issue by FlavienMakesInfographics** [ipython-contrib/jupyter_contrib_nbextensions#792]
Hi,

I came across a behaviour I did not expect, some strange interaction between the Code folding extension and `jupyter nbconvert`.

Let say that a notebook called `test_folding.ipynb` contains the following code
```python
if True:
    print('Whatever')
```

The folding hides `print('Whatever')`, putting a double-headed arrow on the previous line. Now, if the notebook is saved with the code folded and executed through
```
jupyter nbconvert --to notebook --execute test_folding.ipynb --output test_folding_executed.ipynb
```
the new notebook `test_folding_executed.ipynb` does *not* contain anymore `print('Whatever')` but only the double-headed arrow. Here is the resulting cell
```JSON
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Whatever\n"
     ]
    }
   ],
   "source": [
    "if True:↔\n"
   ]
  }
```

Is that an expected behaviour? On a side note, it badly screwed one of my notebook. 

I am using Jupyter version 4.2.0 and IPython version 5.1.0.

Thank you,
F.

**Comment by FlavienMakesInfographics** [comment]
An update on the previous, the following code does produce the proper notebook
```python
import nbformat
from nbconvert.preprocessors import ExecutePreprocessor

with open('test_folding.ipynb') as f:
    nb = nbformat.read(f, as_version=4)

ep = ExecutePreprocessor(timeout=600, kernel_name='python3')
ep.preprocess(nb, {'metadata': {'path': 'notebooks/'}})

with open('test_folding_executed.ipynb', 'wt') as f:
    nbformat.write(nb, f)
```

**Comment by jcb91** [comment]
This is the expected behaviour, to preserve codefolding state in the converted output, courtesy of [the preprocessor `jupyter_contrib_nbextensions.nbconvert_support.CodeFoldingPreprocessor`](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/ca3d33d5cd03657ef70c1afb410f5121a1e1bb28/src/jupyter_contrib_nbextensions/nbconvert_support/pre_codefolding.py), which is enabled when running the install command (see [jupyter_contrib_nbextensions/install.py#L150-L154](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/ca3d33d5cd03657ef70c1afb410f5121a1e1bb28/src/jupyter_contrib_nbextensions/install.py#L150-L154).

In your python snippet, you avoid this behaviour by explicitly calling only a single preprocessor. To not use the codefolding preprocessor, you can edit the `jupyter_nbconvert_config.json` file edited by the install command mentioned above.

Does that make sense?

**Comment by FlavienMakesInfographics** [comment]
From the code, it definitely does. I am a bit surprised by the choice though since I thought that the online converter with `--execute` should have the same behaviour as the "manual" execution cell after cell. But the versatility of the `jupyter nbconvert` may explain the choice.

Thank you for the explanation,
F.

**Comment by jcb91** [comment]
> I am a bit surprised by the choice though since I thought that the online converter with --execute should have the same behaviour as the "manual" execution cell after cell

Yeah, having it enabled by default may not be the best choice, I agree. ~~However, the converter result _is_ the same as executing the notebook manually, since executing manually doesn't affect the codefolding of the input - or have I misunderstood?~~ Apologies, I think I've just understood what you're getting at, in that by executing the codefolding preprocessor _before_ the execute one, it results in invalid python code for the execution. Is that right? If that's the case, then something odd is going on, because I think the default preprocessors (of which execute is one) should be run _before_ any others.

So, to clarify, is your issue that the codefolding preprocessor is just the folding of the cells' inputs in the converted notebook (expected), or also that it is affecting the execute preprocessor (and hence the output of cells in the converted notebook)?

**Comment by juhasch** [comment]
I think the problem is, that when you change a cell while you have some content folded, it can happen that the now stale folding remains. The reason is that codemirror does not fire a folding event and the metadata of the cell does not get updated to reflect that. 

Now the nbconvert preprocessor does not now about these things and might then remove lines that should not be removed.

The solution would be to make sure when making changes to a code cell, or when saving the notebook the folding information gets verified. The second option is probably the way to go, as it should be possible to have codemirror do the work.

**Comment by FlavienMakesInfographics** [comment]
@jcb91 I will try to use the right vocabulary (all of this is very new to me, sorry). The `execute` preprocessor is indeed affected by the `codefolding` one. If you refer my first message, `print('Whatever')` is folded in the genuine notebook. But, when the `execute` is run the cell is wrong with a double-headed arrow in place of the code.

**Comment by jcb91** [comment]
Aha, it all becomes clear! @FlavienMakesInfographics I'm sorry, it's not your vocabulary that's at fault - I misunderstood what you were attempting to do! I missed the fact that you were attempting to convert _to_ a notebook as well as _from_ one.

As far as I can see, then, this issue is the fault of the codefolding preprocessor, the approach of which isn't appropriate for to-notebook conversion. Unlike the nbextension, the preprocessor doesn't actually retain the original source, but replaces folded lines with unicode arrows to represent the fold (see [pre_codefolding.py#L19-L43](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/ca3d33d5cd03657ef70c1afb410f5121a1e1bb28/src/jupyter_contrib_nbextensions/nbconvert_support/pre_codefolding.py#L19-L43) for detail). Although this is fine for static outputs like html, pdf, latex etc, where the fold is simply visual, and the hidden/deleted code unused, it is _an inappropriate behaviour for notebook output_, since the notebook should remain executable (not have code missing), and will have its folding handled in javascript by the nbextension anyway. So, I propose defining a set of output formats for which the preprocessor _should_ be applied (or a set for which it _shouldn't be_, not sure yet which makes more sense), to avoid this situation. @juhasch any thoughts on which output formats need to retain the full source?

**Comment by juhasch** [comment]
I guess the difficulty is that it depends more on the use case than on the output format.
So even if the preprocessor nows what's going on, it will not be the desired result in every case. You might want to export HTML the full cell, just have it folded in the live notebook. 

We could add a configuration option like `NbConvertApp.nbextension.codefolding`, so instead of automatically calling the preprocessor, you have to explicitly enable it:
```
nbconvert --to html mynotebook.ipynb --NbConvertApp.nbextension.codefolding = True
```

**Comment by jcb91** [comment]
> You might want to export HTML the full cell, just have it folded in the live notebook

This is true, but I still can't think of a similar instance for notebook output - I think if you're outputting to notebook, the default behaviour should be to do nothing.

> We could add a configuration option like NbConvertApp.nbextension.codefolding, so instead of automatically calling the preprocessor, you have to explicitly enable it

To be honest, this just sounds like a more complicated way of not enabling the preprocessor in the install step? I guess perhaps seeing it in the `jupyter_nbconvert_config.json` might induce someone to look it up, but it may be more confusing to find it there but not (by default) actually doing anything...

**Comment by juhasch** [comment]
OK, back from travelling.

It certainly makes no sense to clip folded code when converting to a notebook or runnable code. On the other hand, I would like to avoid checking for specific output formats, especially when you can add your own custom exporter.

Adding an option to enable the codefolding preprocessor allows it to be specified at command line. Otherwise you have to manually edit the config file every time. I have found no way to do it otherwise.
As we can install the preprocessor by default without having side effects, it is possible to say
```
jupyter nbconvert --to html --NbConvertApp.codefolding=True mynotebook.ipynb
```
I am happy to hear other options, that make things easier.

**Comment by jcb91** [comment]
Ah, I see what you mean about cli switches - your suggested approach makes
the most sense, I think.

**Comment by jcb91** [comment]
should hopefully be fixed by #794
